packages/node/src/integrations/contextlines.ts

```diff
@@ -0,0 +1,138 @@
+export class ContextLines implements Integration {
+  /**
+   * @inheritDoc
+   */
+  public static id: string = 'ContextLines';
+
+  /**
+   * @inheritDoc
+   */
+  public name: string = ContextLines.id;
+
+  public constructor(private readonly _options: ContextLinesOptions = {}) {}
+
+  /**
+   * @inheritDoc
+   */
+  public setupOnce(addGlobalEventProcessor: (callback: EventProcessor) => void): void {
+    // This is only here to copy frameContextLines from init options if it hasn't
+    // been set via this integrations constructor.
+    //
+    // TODO: Remove on next major!
+    if (this._options.frameContextLines === undefined) {
+      const initOptions = getCurrentHub().getClient<NodeClient>()?.getOptions();
+      // eslint-disable-next-line deprecation/deprecation
+      this._options.frameContextLines = initOptions?.frameContextLines;
+    }
+
+    const contextLines =
+      this._options.frameContextLines !== undefined ? this._options.frameContextLines : DEFAULT_LINES_OF_CONTEXT;
+
+    addGlobalEventProcessor(event => this.addSourceContext(event, contextLines));
+  }
+
+  /** Processes an event and adds context lines */
+  public async addSourceContext(event: Event, contextLines: number): Promise<Event> {
+    const frames = event.exception?.values?.[0].stacktrace?.frames;
+
+    if (frames && contextLines > 0) {
+      const filenames: Set<string> = new Set();
+
+      for (const frame of frames) {
+        if (frame.filename) {
+          filenames.add(frame.filename);
+        }
+      }
+
+      const sourceFiles = await readSourceFiles(filenames);
+
+      for (const frame of frames) {
+        if (frame.filename && sourceFiles[frame.filename]) {
+          try {
+            const lines = (sourceFiles[frame.filename] as string).split('\n');
+            addContextToFrame(lines, frame, contextLines);
+          } catch (e) {
+            // anomaly, being defensive in case
+            // unlikely to ever happen in practice but can definitely happen in theory
+          }
+        }
+      }
+    }
+
+    return event;
+  }
+}
+
+/**
+ * This function reads file contents and caches them in a global LRU cache.
+ *
+ * @param filenames Array of filepaths to read content from.
+ */
+async function readSourceFiles(filenames: Set<string>): Promise<Record<string, string | null>> {
+  const sourceFiles: Record<string, string | null> = {};
+
+  for (const filename of filenames) {
+    const cachedFile = FILE_CONTENT_CACHE.get(filename);
+    // We have a cache hit
+    if (cachedFile !== undefined) {
+      // If stored value is null, it means that we already tried, but couldn't read the content of the file. Skip.
+      if (cachedFile === null) {
+        continue;
+      }
+
+      // Otherwise content is there, so reuse cached value.
+      sourceFiles[filename] = cachedFile;
+      continue;
+    }
+
+    let content: string | null = null;
+    try {
+      content = await readTextFileAsync(filename);
```

Comment:
Can we make this sync for now? And come back later to evaluate if we should use promises? We used `SyncPromise` here previously.

Reply:
The async code we are thinking about removing later is in the `Backend` + `Parser` which appears like it was only there to cater for node.js code context loading which this PR moves.

The previous code might use `SyncPromise` but it [also calls the async `readFile` function](https://github.com/getsentry/sentry-javascript/blob/1bf988322a6adb931905305747f4fafb742002b1/packages/node/src/parsers.ts#L108-L120) so my understanding is that previously this code was already async.

This PR moves the context loading into the event processor pipeline and my understanding is that async there is perfectly safe/ok.

Reply:
Ah that’s on me getting things mixed up here, and why I should maybe reduce jumping between dozens of topics during my day 😅

sounds good, let’s ship this then